README.md
<!-- badges: start -->
[![R-CMD-check-bioc](https://github.com/adeschen/BasalClassicalHeatmap/actions/workflows/check-bioc.yaml/badge.svg)](https://github.com/adeschen/BasalClassicalHeatmap/actions/workflows/check-bioc.yaml)
[![codecov](https://codecov.io/gh/adeschen/BasalClassicalHeatmap/branch/main/graph/badge.svg?token=XVKU8S1E2R)](https://codecov.io/gh/adeschen/BasalClassicalHeatmap)
[![License: Artistic-2.0](https://img.shields.io/badge/License-Artistic%202.0-0298c3.svg)](https://opensource.org/licenses/Artistic-2.0)
<!-- badges: end -->

# BasalClassicalHeatmap

The __BasalClassicalHeatmap__ package enables Basal-like/Classical 
classification of pancreatic tumor 
samples by generating heatmaps using transcriptomic data. 

Three different lists of Basal-like/Classical genes are 
available in the __BasalClassicalHeatmap__ package: the original list 
from Moffitt et al. 2015; the 
updated list from Miyabayashi et al. 2020 and the list from 
Tiriac et al. 2018 that has been developed for pancreatic ductal 
adenocarcinoma (PDAC) organoids.


## Authors ##

[Astrid Desch&ecirc;nes](http://ca.linkedin.com/in/astriddeschenes "Astrid Desch&ecirc;nes"),
Daniel King and
David A. Tuveson


## Citation ##

If you use this package for a publication, we would ask you 
to cite the following:

> Deschênes A, King D and Tuveson DA. BasalClassicalHeatmap: Heatmap using basal-like and classical gene list and RNA-seq expression data. R package version 0.0.3, https://adeschen.github.io/BasalClassicalHeatmap/ 
  
You can also replace the latest by the Research Resource Identifier __(RRID)__, as 
example:

>   BasalClassicalHeatmap version 0.0.3, RRID: SCR_022949

If you use one of the available Basal-like/Classical gene list, the appropriate
paper should be referenced.

For Moffitt et al. 2015

> Moffitt RA, Marayati R, Flate EL et al. Virtual microdissection identifies distinct tumor- and stroma-specific subtypes of pancreatic ductal adenocarcinoma. Nat Genet. 2015 Oct;47(10):1168-78. doi: 10.1038/ng.3398. Epub 2015 Sep 7. PMID: 26343385; PMCID: PMC4912058.

For Tiriac et al. 2018

> Tiriac H, Belleau P, Engle DD et al. Organoid Profiling Identifies Common Responders to Chemotherapy in Pancreatic Cancer. Cancer Discov. 2018 Sep;8(9):1112-1129. doi: 10.1158/2159-8290.CD-18-0349. Epub 2018 May 31. PMID: 29853643; PMCID: PMC6125219.

For Miyabayashi et al. 2020

> Miyabayashi K, Baker LA, Deschênes A et al. Intraductal Transplantation Models of Human Pancreatic Ductal Adenocarcinoma Reveal Progressive Transition of Molecular Subtypes. Cancer Discov. 2020 Oct;10(10):1566-1589. doi: 10.1158/2159-8290.CD-20-0133. Epub 2020 Jul 23. PMID: 32703770; PMCID: PMC7664990.


## Documentation ##

[BasalClassicalHeatmap Website](https://adeschen.github.io/BasalClassicalHeatmap/)

[BasalClassicalHeatmap Get Started](https://adeschen.github.io/BasalClassicalHeatmap/articles/BasalClassicalHeatmap.html)

## Installation ##

To install the latest version accessible using the [devtools](https://cran.r-project.org/web/packages/devtools/index.html) 
package:

     ## Load required package
     library(devtools)

     ## Install the latest version of BasalClassicalHeatmap
     devtools::install_github('adeschen/BasalClassicalHeatmap')



## Maintainer

[Astrid Desch&ecirc;nes](https://github.com/adeschen/ "Astrid Desch&ecirc;nes")


## Bugs/Feature requests ##

[Please contact us](https://github.com/adeschen/BasalClassicalHeatmap/issues) for bug fixes or with feature requests. 

Thanks!



NEWS
CHANGES IN VERSION 0.0.4
------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

   o The 'log_data' parameter that enable/disable log transformation of the data has been added to the 'create_createHeatmap' function
  

CHANGES IN VERSION 0.0.3
------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

   o The vignette has been upgraded
   o The citations for the basal-like/classical gene lists have been added
   

CHANGES IN VERSION 0.0.2
------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

   o The vignette has been upgraded

MODIFICATIONS

   o The createHeatmap() function has a new 'heatmap_legend_param' parameter
   o New unit tests

tests/testthat.R
library(testthat)
library(BasalClassicalHeatmap)

## Run all unit tests
test_check("BasalClassicalHeatmap")

tests/testthat/test-heatmap.R
### Unit tests for Heatmap.R functions

library(BasalClassicalHeatmap)


#############################################################################
### Tests createHeatmap() results
#############################################################################

context("createHeatmap() results")

test_that("createHeatmap() must return error when gene_column is a number", {
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                        "Sample2"=sample(1:100, size=10, replace=T),
                        "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    error_message <- paste0("The \'gene_column\' parameter must be a ", 
                                "character string.")
    
    expect_error(createHeatmap(rna_data=demo, gene_list=demo$GENE, 
                    gene_column=33), error_message)
})


test_that("createHeatmap() must return error when gene_column is not in rna_demo column names", {
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                       "Sample2"=sample(1:100, size=10, replace=T),
                       "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    error_message <- paste0("The \'gene_column\' parameter must correspond to ",
        "the name of a column in the \'rna_data\' parameter.")
    
    expect_error(createHeatmap(rna_data=demo, gene_list=demo$GENE, 
        gene_column="TEST"), error_message)
})


test_that("createHeatmap() must return error when show_column_dend is a number", {
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                        "Sample2"=sample(1:100, size=10, replace=T),
                        "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    error_message <- paste0("The \'show_column_dend\' parameter must be a ", 
                                "logical (TRUE OR FALSE).")
    
    expect_error(createHeatmap(rna_data=demo, gene_list=demo$GENE, 
        gene_column="GENE", show_column_dend=22), error_message, fixed=TRUE)
})


test_that("createHeatmap() must return error when show_row_dend is a number", {
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                        "Sample2"=sample(1:100, size=10, replace=T),
                        "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    error_message <- paste0("The \'show_row_dend\' parameter must be a ", 
                                "logical (TRUE OR FALSE).")
    
    expect_error(createHeatmap(rna_data=demo, gene_list=demo$GENE, 
        gene_column="GENE", show_row_dend=22), error_message, fixed=TRUE)
})


test_that("createHeatmap() must return error when cluster_rows is a number", {
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                        "Sample2"=sample(1:100, size=10, replace=T),
                        "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    error_message <- paste0("The \'cluster_rows\' parameter must be a ", 
                                "logical (TRUE OR FALSE).")
    
    expect_error(createHeatmap(rna_data=demo, gene_list=demo$GENE, 
        gene_column="GENE", cluster_rows=22), error_message, fixed=TRUE)
})


test_that("createHeatmap() must return error when cluster_columns is a number", {
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                        "Sample2"=sample(1:100, size=10, replace=T),
                        "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    error_message <- paste0("The \'cluster_columns\' parameter must be a ", 
                                "logical (TRUE OR FALSE).")
    
    expect_error(createHeatmap(rna_data=demo, gene_list=demo$GENE, 
        gene_column="GENE", cluster_columns=21), error_message, fixed=TRUE)
})


test_that("createHeatmap() must return error when log_data is a number", {
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                       "Sample2"=sample(1:100, size=10, replace=T),
                       "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    error_message <- paste0("The \'log_data\' parameter must be a ", 
                                "logical (TRUE OR FALSE).")
    
    expect_error(createHeatmap(rna_data=demo, gene_list=demo$GENE, 
        gene_column="GENE", cluster_columns=TRUE, log_data=11), 
        error_message, fixed=TRUE)
})

tests/testthat/test-heatmap_internal.R
### Unit tests for Heatmap_internal.R functions

library(BasalClassicalHeatmap)


#############################################################################
### Tests validateCreateHeatmap() results
#############################################################################

context("validateCreateHeatmap() results")

test_that("validateCreateHeatmap() must return expected results when all parameters are ok", {
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                       "Sample2"=sample(1:100, size=10, replace=T),
                       "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    result <- BasalClassicalHeatmap:::validateCreateHeatmap(rna_data=demo, 
        gene_list=demo$GENE, gene_column="GENE", name="Expression",
        show_column_dend=TRUE, show_row_dend=TRUE, cluster_columns=TRUE, 
        cluster_rows=TRUE, log_data=FALSE)
    
    expect_equal(result, 0L)
})


#############################################################################
### Tests prepareExpression() results
#############################################################################

context("prepareExpression() results")

test_that("prepareExpression() must return expected results test 01", {
    
    set.seed(121)
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                       "Sample2"=sample(1:100, size=10, replace=T),
                       "Sample3"=sample(1:100, size=10, replace=T),
                       "Sample4"=sample(1:100, size=10, replace=T),
                       "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    gene_list <- data.frame(SYMBOL=c("GENE_2", "GENE_4", "GENE_5"), 
                    Class=c("Heavy", "Heavy", "Light"))
    
    
    results <- prepareExpression(rna_data=demo, gene_list=gene_list, 
                                    gene_column="GENE", log_data=TRUE) 
    
    expected <- list()
    expected[["DATA"]] <- matrix(data=c(-0.647362460105595, 0.953718090017069, 
                            -1.491481935002304, -0.493042823565605, 
                            -1.404530776770527, 0.605386682277673,
                            -0.348374658630693, 0.315125523627209,
                            0.352476648582310, 1.488779942301896,
                            0.135687163126249, 0.533618604142320), 
                            byrow = FALSE, nrow = 3)
    rownames(expected[["DATA"]]) <- c("GENE_2", "GENE_4", "GENE_5")
    colnames(expected[["DATA"]]) <- c("Sample1", "Sample2", "Sample3", 
                                        "Sample4")
    
    expected[["ROW_INFO"]] <- gene_list
    rownames(expected[["ROW_INFO"]]) <- c("GENE_2", "GENE_4", "GENE_5")
        
    expect_equal(results, expected) 
})


test_that("prepareExpression() must return expected results test 02", {
    
    set.seed(121)
    
    demo <- data.frame("Sample1"=sample(1:100, size=10, replace=T),
                       "Sample2"=sample(1:100, size=10, replace=T),
                       "Sample3"=sample(1:100, size=10, replace=T),
                       "Sample4"=sample(1:100, size=10, replace=T),
                       "GENE"=paste0("GENE_", 1:10), stringsAsFactors=FALSE)
    
    gene_list <- data.frame(SYMBOL=c("GENE_2", "GENE_4", "GENE_5"), 
                            Class=c("Heavy", "Heavy", "Light"))
    
    
    results <- prepareExpression(rna_data=demo, gene_list=gene_list, 
                                 gene_column="GENE", log_data=FALSE) 
    
    expected <- list()
    expected[["DATA"]] <- matrix(data=c(-0.593446554394425, 1.265636487594504, 
                                        -1.440717520307526, -0.498495105691317, 
                                        -1.169023015259046, 0.773437826691409,
                                        -0.403543656988209, 0.067629430634821,
                                        0.106158133075291, 1.495485317073950,
                                        -0.164242902970279, 0.561121560540826), 
                                 byrow = FALSE, nrow = 3)
    rownames(expected[["DATA"]]) <- c("GENE_2", "GENE_4", "GENE_5")
    colnames(expected[["DATA"]]) <- c("Sample1", "Sample2", "Sample3", 
                                      "Sample4")
    
    expected[["ROW_INFO"]] <- gene_list
    rownames(expected[["ROW_INFO"]]) <- c("GENE_2", "GENE_4", "GENE_5")
    
    expect_equal(results, expected) 
})

R/Heatmap.R
#' @title Generate a heatmap based on transcriptomic data for a 
#' basal-like/classical gene list
#'
#' @description The function selects the expression values related to the 
#' gene list passed in argument. Then, the function adds one to the expression 
#' values before log transforming the values. Then, the function normalizes 
#' the data so that each row has mean 0 and variance 1. Finally, the function 
#' generates a heatmap using input parameters.
#'
#' @param gene_list a \code{data.frame} containing the gene list used for the
#' basal-like/classical classification. The \code{data.frame} must contain 
#' two columns: \code{SYMBOL} and \code{Class}. The \code{SYMBOL} contains the 
#' gene symbol while the \code{Class} column contains its classification. The 
#' classification can be \code{Classical} or \code{Basal-like}. 
#' 
#' @param rna_data  \code{data.frame} containing the expression level for the 
#' samples. Each column must represent a sample. An extra column for the 
#' gene name must be present. The name of this column must correspond to the 
#' parameter \code{gene_column}.
#' 
#' @param gene_column a \code{character} string representing the name of the  
#' column in the \code{rna_data} parameter that contains the gene names. 
#' Default: \code{GENE}.
#'
#' @param name a \code{character} string representing the title of the 
#' heatmap legend. Default: \code{"Expression"}.
#' 
#' @param clustering_distance_rows a \code{character} string representing 
#' the method uses to cluster the rows of the heatmap. 
#' The available choices are: 
#' "euclidean", "maximum", "manhattan", "canberra", 
#' "binary", "minkowski", "pearson", "spearman" and "kendall". 
#'  Default: \code{"euclidean"}.
#'
#' @param clustering_distance_columns a \code{character} string representing 
#' the method uses to cluster the columns of the heatmap.
#' The available choices are: 
#' "euclidean", "maximum", "manhattan", "canberra", 
#' "binary", "minkowski", "pearson", "spearman" and "kendall". 
#' Default: \code{"euclidean"}.
#' 
#' @param show_column_dend a code{logical} indicating if the column 
#' dendrogram should be shown. Default: \code{TRUE}.
#' 
#' @param show_row_dend a code{logical} indicating if the row  
#' dendrogram should be shown. Default: \code{TRUE}.
#'
#' @param cluster_columns a code{logical} indicating if the columns   
#' should be clustered. Default: \code{TRUE}.
#' 
#' @param cluster_rows a code{logical} indicating if the rows   
#' should be clustered. Default: \code{TRUE}.
#' 
#' @param heatmap_legend_param an object of class '\code{gpar}'. Default: 
#' \code{gpar(title_gp=gpar(col="black", fontsize=11, fontface="bold"))}.
#' 
#' @param log_data a \code{logical} indicating if the data should be log 
#' transformed. If so, the data will be log2(value + 1) transformed. 
#' Default: \code{TRUE}.
#' 
#' @param \ldots further arguments passed to ComplexHeatmap::Heatmap() function.
#'
#' @return an object of class {ComplexHeatmap} that represent a heatmap based 
#' on transcriptomic data for a specific basal-like/classical gene list.
#'
#' @examples
#'
#' ## Load Moffitt et al 2015 gene list
#' data("Moffitt_2015")
#'
#' ## Load demo PDAC TCGA expression data
#' data("demo_PDAC_TCGA")
#'
#' ## Create heatmap with default parameters
#' ## The column in the rna_data data.frame is the default name "GENE"
#' createHeatmap(gene_list=Moffitt_2015, rna_data=demo_PDAC_TCGA)
#' 
#' @seealso
#' 
#' The default method ComplexHeatmap::Heatmap().
#' 
#' 
#' @author Astrid Deschênes
#' @importFrom ComplexHeatmap Heatmap HeatmapAnnotation 
#' @importFrom grid gpar
#' @encoding UTF-8
#' @export
createHeatmap <- function(gene_list, rna_data, gene_column="GENE",
    name="Expression",
    clustering_distance_rows=c("euclidean", "maximum", "manhattan", "canberra", 
    "binary", "minkowski", "pearson", "spearman", "kendall"), 
    clustering_distance_columns=c("euclidean", "maximum", "manhattan", 
    "canberra", "binary", "minkowski", "pearson", "spearman", "kendall"),
    show_column_dend=TRUE, show_row_dend=TRUE, 
    cluster_columns=TRUE, cluster_rows=TRUE, log_data=TRUE,
    heatmap_legend_param=gpar(title_gp=gpar(col="black", fontsize=11, 
    fontface="bold")), ...) {
    
    ## Validate parameters
    validateCreateHeatmap(gene_list=gene_list, rna_data=rna_data, 
        gene_column=gene_column, name=name, show_column_dend=show_column_dend, 
        show_row_dend=show_row_dend, cluster_columns=cluster_columns, 
        cluster_rows=cluster_rows, log_data=log_data)
    
    ## Select clustering rows method to be used
    clustering_distance_rows <- match.arg(clustering_distance_rows)
    
    ## Select clustering column method to be used
    clustering_distance_columns <- match.arg(clustering_distance_columns)
    
    cleanData <- prepareExpression(rna_data, gene_list, gene_column="GENE",
                                   log_data=log_data)
    
    row_ha <- HeatmapAnnotation(Subtype=cleanData[["ROW_INFO"]]$Class,
                    col=list(Subtype=c("Classical"="darkviolet", 
                        "Basal-like"="darkorange")),
                    annotation_name_gp=gpar(fontsize = 12),
                    annotation_legend_param=list(Subtype=list(border="black", 
                        direction="vertical", title_gp=gpar(fontsize=12, 
                        fontface="bold"))),
                    annotation_name_side=list(Subtype="bottom"), which="row")
    
    gg <- Heatmap(matrix=cleanData[["DATA"]], right_annotation=row_ha,
            name=name, heatmap_legend_param=heatmap_legend_param, 
            clustering_distance_rows=clustering_distance_rows, 
            clustering_distance_columns=clustering_distance_columns, 
            show_column_dend=show_column_dend,
            cluster_columns=cluster_columns, ...)
    
    return(gg)
}

R/Heatmap_internal.R
#' @title Prepare the expression data by selecting basal-like/classical genes
#' and running normalization
#'
#' @description The function selects the expression values related to the 
#' gene list passed in argument. Then, the function adds one to the expression 
#' values before log transforming the values. Then, the function normalizes 
#' the data so that each row has mean 0 and variance 1.
#'
#' @param rna_data a \code{data.frame} containing the expression values for 
#' the studied samples. There must be one column per sample and 
#' an extra column for the gene names. The name of the column containing the 
#' gene names is passed in the \code{gene_column} parameter.
#' 
#' @param gene_list a \code{data.frame} containing the gene list used for the
#' basal-like/classical classification.
#'
#' @param gene_column a \code{character} string representing the name of the  
#' column in the \code{rna_data} parameter that contains the gene names. 
#' 
#' @param log_data a \code{logical} indicating if the data should be log 
#' transformed. If so, the data will be log2(value + 1) transformed. 
#' Default: \code{TRUE}.
#'
#' @return a \code{list} containing 2 entries: \itemize{
#' \item{DATA} { a \code{data.frame} containing the normalized expression 
#' values for the genes that are part of the classical/basal-like gene list. 
#' The gene names are used as row names and the gene order fits the order 
#' of the gene names in the \code{ROW_INFO} \code{data.frame}.}
#' \item{ROW_INFO} {a \code{data.frame} containing information about the genes 
#' from basal-like/classical classification that are present in the 
#' \code{DATA} \code{data.frame}. The gene names are used as row names and the 
#' gene order fits the order of the gene names in 
#' the \code{DATA} \code{data.frame}.}
#' }
#'
#' @examples
#'
#' ## Load Moffitt et al 2015 gene list
#' data("Moffitt_2015")
#'
#' ## Load demo PDAC TCGA expression data
#' data("demo_PDAC_TCGA")
#'
#' ## Prepare the data for the heatmap creation step
#' result <- BasalClassicalHeatmap:::prepareExpression(rna_data=demo_PDAC_TCGA, 
#'     gene_list=Moffitt_2015, gene_column="GENE", log_data=TRUE)
#'
#' @author Astrid Deschênes
#' @importFrom som normalize
#' @encoding UTF-8
#' @keywords internal
prepareExpression <- function(rna_data, gene_list, gene_column="GENE", 
                              log_data=TRUE) {
    
    ## Create expression data frame containing only genes in list
    clean_data <- rna_data[rna_data[[gene_column]] %in% gene_list$SYMBOL,]
    clean_data <- clean_data[!duplicated(clean_data[[gene_column]]), ]
    rownames(clean_data) <- clean_data[[gene_column]]
    clean_data <- clean_data[, c(colnames(clean_data) != gene_column)]
    
    ## Create data frame with row information
    rowInfo <- gene_list[gene_list$SYMBOL %in% rownames(clean_data),]
    rownames(rowInfo) <- rowInfo$SYMBOL
    rowInfo <- rowInfo[rownames(clean_data),]
    
    ## Normalize data
    if (log_data) {
        temp <- colnames(clean_data)
        normRes <- log2(clean_data + 1)
        normResNorm <- normalize(normRes, byrow=TRUE)
        colnames(normResNorm) <- temp
    } else {
        temp <- colnames(clean_data)
        normResNorm <- normalize(clean_data, byrow=TRUE)
        colnames(normResNorm) <- temp
    }
    
    listR <- list()
    listR[["DATA"]] <- normResNorm
    listR[["ROW_INFO"]] <- rowInfo
        
    return(listR)
}


#' @title Validate the parameters for createHeatmap() function
#'
#' @description This function validates the parameters input in the 
#' createHeatmap() function.
#'
#' @param gene_list a \code{data.frame} containing the gene list used for the
#' basal-like/classical classification. The \code{data.frame} must contain 
#' two columns: \code{SYMBOL} and \code{Class}. The \code{SYMBOL} contains the 
#' gene symbol while the \code{Class} column contains its classification. The 
#' classification can be \code{Classical} or \code{Basal-like}. 
#' 
#' @param rna_data  \code{data.frame} containing the expression level for the 
#' samples. Each column must represent a sample. An extra column for the 
#' gene name must be present. The name of this column must correspond to the 
#' parameter \code{gene_column}.
#' 
#' @param gene_column a \code{character} string representing the name of the  
#' column in the \code{rna_data} parameter that contains the gene names. 
#' Default: \code{GENE}.
#'
#' @param name a \code{character} string representing the title of the 
#' heatmap legend.
#' 
#' @param show_column_dend a code{logical} indicating if the column 
#' dendrogram should be shown. 
#' 
#' @param show_row_dend a code{logical} indicating if the row  
#' dendrogram should be shown. 
#'
#' @param cluster_columns a code{logical} indicating if the columns   
#' should be clustered. 
#' 
#' @param cluster_rows a code{logical} indicating if the rows   
#' should be clustered. 
#' 
#' @param log_data a code{logical} indicating if the expression data    
#' should be log transformed. 
#'
#' @return \code{0L} when the function is successful.
#'
#' @examples
#'
#' ## Load Moffitt et al 2015 gene list
#' data("Moffitt_2015")
#'
#'
#' ## Load TCGA PDAC dataset
#' data("demo_PDAC_TCGA")
#' 
#' ## TODO
#' BasalClassicalHeatmap:::validateCreateHeatmap(gene_list=Moffitt_2015,
#'     rna_data=demo_PDAC_TCGA, gene_column="GENE", name="Expression", 
#'     show_column_dend=TRUE, show_row_dend=TRUE, cluster_columns=TRUE,
#'     cluster_rows=TRUE, log_data=FALSE)
#' 
#' 
#' @author Astrid Deschênes
#' @encoding UTF-8
#' @keywords internal
validateCreateHeatmap <- function(gene_list, rna_data, gene_column,
    name, show_column_dend, show_row_dend, cluster_columns, cluster_rows, 
    log_data) {
    
    if (!is.character(gene_column)) {
        stop("The \'gene_column\' parameter must be a character string.")
    }
    
    if (!gene_column %in% colnames(rna_data)) {
        stop("The \'gene_column\' parameter must correspond to the name of ",
                "a column in the \'rna_data\' parameter.")
    }
    
    if (!is.logical(show_column_dend)) {
        stop("The \'show_column_dend\' parameter must be a logical ", 
                "(TRUE OR FALSE).")
    }
    
    if (!is.logical(show_row_dend)) {
        stop("The \'show_row_dend\' parameter must be a logical ", 
                "(TRUE OR FALSE).")
    }
    
    if (!is.logical(cluster_rows)) {
        stop("The \'cluster_rows\' parameter must be a logical ", 
                "(TRUE OR FALSE).")
    }
    
    if (!is.logical(cluster_columns)) {
        stop("The \'cluster_columns\' parameter must be a logical ", 
                "(TRUE OR FALSE).")
    }
    
    if (!is.logical(log_data)) {
        stop("The \'log_data\' parameter must be a logical ", 
             "(TRUE OR FALSE).")
    }
    
    return(0L)
}
    

R/BasalClassicalHeatmap.R
#' BasalClassicalHeatmap: TODO
#'
#' To learn more about \pkg{BasalClassicalHeatmap} package see:
#' \url{https://github.com/adeschen/BasalClassicalHeatmap/wiki}
#'
#' @docType package
#'
#' @name BasalClassicalHeatmap-package
#'
#' @aliases BasalClassicalHeatmap-package BasalClassicalHeatmap
#'
#' @author
#'
#' Maintainer:
#' Astrid Deschênes <adeschen@@hotmail.com>
#'
#' @seealso
#' \itemize{
#' \item \code{\link{createHeatmap}} {for creating a heating using a 
#' gene list and RNA-seq expression data.}
#' }
#' @encoding UTF-8
#' @keywords package
NULL


#' The Basal-like and Classical gene list used in Tiriac et al 2018.
#'
#' @name Tiriac_2018
#'
#' @docType data
#'
#' @aliases Tiriac_2018
#'
#' @format a \code{data.frame} with the following columns:
#' \itemize{
#' \item \code{SYMBOL} { a \code{character} string representing the
#' symbol of the gene. }
#' \item \code{Class} { a \code{character} string representing the class 
#' (Basal-like or Classical) associated to the gene. }
#' }
#'
#' @seealso
#' \itemize{
#' \item \code{\link{createHeatmap}} { for creating a heating using a 
#' gene list and RNA-seq expression data.}
#' }
#'
#' @usage data(Tiriac_2018)
#'
#' @keywords datasets
#'
#' @references 
#' 
#' Tiriac H, Belleau P, Engle DD, Plenker D, Deschênes A, Somerville TDD, 
#' Froeling FEM, Burkhart RA, Denroche RE, Jang GH, Miyabayashi K, Young CM, 
#' Patel H, Ma M, LaComb JF, Palmaira RLD, Javed AA, Huynh JC, Johnson M, 
#' Arora K, Robine N, Shah M, Sanghvi R, Goetz AB, Lowder CY, Martello L, 
#' Driehuis E, LeComte N, Askan G, Iacobuzio-Donahue CA, Clevers H, Wood LD, 
#' Hruban RH, Thompson E, Aguirre AJ, Wolpin BM, Sasson A, Kim J, Wu M, 
#' Bucobo JC, Allen P, Sejpal DV, Nealon W, Sullivan JD, Winter JM, 
#' Gimotty PA, Grem JL, DiMaio DJ, Buscaglia JM, Grandgenett PM, Brody JR, 
#' Hollingsworth MA, O'Kane GM, Notta F, Kim E, Crawford JM, Devoe C, Ocean A, 
#' Wolfgang CL, Yu KH, Li E, Vakoc CR, Hubert B, Fischer SE, Wilson JM, 
#' Moffitt R, Knox J, Krasnitz A, Gallinger S, Tuveson DA. 
#' Organoid Profiling Identifies Common Responders to Chemotherapy in 
#' Pancreatic Cancer. Cancer Discov. 2018 Sep;8(9):1112-1129. 
#' doi: 10.1158/2159-8290.CD-18-0349. 
#' Epub 2018 May 31. PMID: 29853643; PMCID: PMC6125219.
#' 
#' @examples
#'
#' ## Load Tiriac et al 2018 gene list used for 
#' ## Basal-like/Classical classification
#' data(Tiriac_2018)
#'
#' ## The dataset contains 2 columns
#' ## The first column has the gene symbols
#' ## The second column the class assignation (Basal-like or Classical)
#' head(Tiriac_2018)
#'
#'
NULL


#' The Basal-like and Classical gene list used in Moffitt et al 2015.
#'
#' @name Moffitt_2015
#'
#' @docType data
#'
#' @aliases Moffitt_2015
#'
#' @format a \code{data.frame} with the following columns:
#' \itemize{
#' \item \code{SYMBOL} { a \code{character} string representing the
#' symbol of the gene. }
#' \item \code{Class} { a \code{character} string representing the class 
#' (Basal-like or Classical) associated to the gene. }
#' }
#'
#' @seealso
#' \itemize{
#' \item \code{\link{createHeatmap}} { for creating a heating using a 
#' gene list and RNA-seq expression data.}
#' }
#'
#' @usage data(Moffitt_2015)
#'
#' @keywords datasets
#'
#' @references 
#' 
#' Moffitt RA, Marayati R, Flate EL, Volmar KE, Loeza SG, Hoadley KA, 
#' Rashid NU, Williams LA, Eaton SC, Chung AH, Smyla JK, Anderson JM, 
#' Kim HJ, Bentrem DJ, Talamonti MS, Iacobuzio-Donahue CA, Hollingsworth MA, 
#' Yeh JJ. Virtual microdissection identifies distinct tumor- and 
#' stroma-specific subtypes of pancreatic ductal adenocarcinoma. 
#' Nat Genet. 2015 Oct;47(10):1168-78. 
#' doi: 10.1038/ng.3398. Epub 2015 Sep 7. PMID: 26343385; PMCID: PMC4912058.
#' 
#' @examples
#'
#' ## Load Moffitt et al 2015 gene list used for 
#' ## Basal-like/Classical classification
#' data(Moffitt_2015)
#'
#' ## The dataset contains 2 columns
#' ## The first column has the gene symbols
#' ## The second column the class assignation (Basal-like or Classical)
#' head(Moffitt_2015)
#'
NULL


#' The Basal-like and Classical gene list used in Miyabayashi et al 2020.
#'
#' @name Miyabayashi_2020
#'
#' @docType data
#'
#' @aliases Miyabayashi_2020
#'
#' @format a \code{data.frame} with the following columns:
#' \itemize{
#' \item \code{SYMBOL} { a \code{character} string representing the
#' symbol of the gene. }
#' \item \code{Class} { a \code{character} string representing the class 
#' (Basal-like or Classical) associated to the gene. }
#' }
#'
#' @seealso
#' \itemize{
#' \item \code{\link{createHeatmap}} { for creating a heating using a 
#' gene list and RNA-seq expression data.}
#' }
#'
#' @usage data(Miyabayashi_2020)
#'
#' @keywords datasets
#'
#' @references 
#' 
#' Miyabayashi K, Baker LA, Deschênes A, Traub B, Caligiuri G, Plenker D, 
#' Alagesan B, Belleau P, Li S, Kendall J, Jang GH, Kawaguchi RK, 
#' Somerville TDD, Tiriac H, Hwang CI, Burkhart RA, Roberts NJ, Wood LD, 
#' Hruban RH, Gillis J, Krasnitz A, Vakoc CR, Wigler M, Notta F, 
#' Gallinger S, Park Y, Tuveson DA. Intraductal Transplantation Models of 
#' Human Pancreatic Ductal Adenocarcinoma Reveal Progressive Transition of 
#' Molecular Subtypes. Cancer Discov. 2020 Oct;10(10):1566-1589. 
#' doi: 10.1158/2159-8290.CD-20-0133. 
#' Epub 2020 Jul 23. PMID: 32703770; PMCID: PMC7664990.
#' 
#' @examples
#'
#' ## Load Miyabayashi et al 2020 gene list used for 
#' ## Basal-like/Classical classification
#' data(Miyabayashi_2020)
#'
#' ## The dataset contains 2 columns
#' ## The first column has the gene symbols
#' ## The second column the class assignation (Basal-like or Classical)
#' head(Miyabayashi_2020)
#'
NULL



#' TODO
#'
#' @name demo_PDAC_TCGA
#'
#' @docType data
#'
#' @aliases demo_PDAC_TCGA
#'
#' @format a \code{data.frame} containing normalized expected counts for genes 
#' associated to Basal-like/Classical classification. The rows are containing 
#' normalized expected counts. The \code{data.frame} 
#' contains one column per sample in addition to a column called 'GENE' that 
#' contains the name of the genes represented in each row.
#'
#' @seealso
#' \itemize{
#' \item \code{\link{createHeatmap}} { for creating a heating using a 
#' gene list and RNA-seq expression data.}
#' }
#'
#' @usage data(demo_PDAC_TCGA)
#'
#' @keywords datasets
#' 
#' @references
#' 
#' The TCGA Research Network: https://www.cancer.gov/tcga
#'
#' @examples
#'
#' ## Load Miyabayashi et al 2020 gene list used for 
#' ## Basal-like/Classical classification
#' data(Miyabayashi_2020)
#'
#' ## Load the expression dataset
#' data(demo_PDAC_TCGA)
#'
#' ## Create a heatmap using the Miyabayashi et al 2020 gene list
#' createHeatmap(gene_list=Miyabayashi_2020, rna_data=demo_PDAC_TCGA, 
#'      gene_column="GENE")
#'      
NULL